### PR TITLE
Update to the latest c-code template

### LINKED ISSUE #153: Python 2: OxBTrees allow types as keys; Python 3 does not
Consider:
```python
class Old:
    pass

class New(object):
    pass

from BTrees import family64
bt = family64.OO.BTree()
```

On Python 3, both `Old` and `New`, which are types, cannot go in the tree because they have default comparison:
```pycon
>>> bt[Old] = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Object has default comparison
>>> bt[New] = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Object has default comparison
```

Under Python 2, however, the default comparison is only detected for old-style classes; new style `type` apparently doesn't have default comparison that we can detect:
```pycon
>>> bt[Old] = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Object has default comparison
>>> bt[New] = 1
```

Type ordering in Python 2 [is based on address](https://github.com/python/cpython/blob/8d21aa21f2cbc6d50aab3f420bb23be1d081dac4/Objects/typeobject.c#L669-L679), making it not suitable for use in a BTree. It emits a warning if you run Python with the `-3` argument.

Can we try harder (or just special case) to find that `type` objects shouldn't be used as keys?

### PR DESCRIPTION
**Pull request by mgedmin**
Supersedes and closes #179.  (I wonder if GitHub will parse that?)

I had trouble running the tests locally (no Python 3.5, and my 3.6 build seems to be broken), so let's try it on CI.

### CODE REVIEW

**Review by icemac — APPROVED** [review]
LGTM.

**Comment by mgedmin** [comment]
https://github.com/zopefoundation/meta/pull/153#issuecomment-1239907333 makes me want to add a regression test for this.

**Comment by mgedmin** [comment]
I verified that the regression test catches the problem by doing

```
$ export CFLAGS=-Ofast
$ tox -re py38 -- -pvc -m test_compile_flags -t 'test_no_fast_math_optimization [(].*[.]TestFloatingPoint[)]'
...
Failure in test test_no_fast_math_optimization (BTrees.tests.test_compile_flags.TestFloatingPoint)
Traceback (most recent call last):
  File "/home/mg/opt/python38/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/home/mg/opt/python38/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/home/mg/opt/python38/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/mg/src/zopefoundation/BTrees/src/BTrees/tests/test_compile_flags.py", line 29, in test_no_fast_math_optimization
    self.assertNotEqual(smallest_subnormal, 0.0)
  File "/home/mg/opt/python38/lib/python3.8/unittest/case.py", line 921, in assertNotEqual
    raise self.failureException(msg)
AssertionError: 0.0 == 0.0
...
```

**Comment by mgedmin** [comment]
Hey, `.manylinux-install.sh` has an `export CFLAGS="-Ofast $CFLAGS"`.  That one comes from `.meta.toml` in this repo, and it needs to be updated as well.

**Comment by mgedmin** [comment]
The test indicates that @srett was right and -fno-unsafe-math-optimizations does **not** override -Ofast.

-fno-fast-math doesn't either.

I think we're going to have to switch to -O3 or something like that.

**Comment by mgedmin** [comment]
Let's try this again once https://github.com/zopefoundation/meta/pull/155 is merged.

**Comment by mgedmin** [comment]
The new test is failing and I have no idea why.  There's no sign of -Ofast nor -ffast-math in the build log.  Local testing with -O3 tox -r passes for me.

Somebody help?

**Comment by d-maurer** [comment]
> The new test is failing and I have no idea why. There's no sign of -Ofast nor -ffast-math in the build log. Local testing with -O3 tox -r passes for me.
> 
> Somebody help?

`BTrees` depends on `persistent`, i.e. another extension implemented in "C". Maybe, this comes from the `pip` cache and has been compiled with `-Ofast`.

**Comment by mgedmin** [comment]
> BTrees depends on persistent, i.e. another extension implemented in "C". Maybe, this comes from the pip cache and has been compiled with -Ofast.

This sounds very plausible!

I see `persistent` also has CFLAGS: -Ofast in the GitHub Actions workflow, but not in .manylinux-install.sh.  I think the binary wheels on PyPI don't have the math trap, or `persistent` would have been mentioned in https://moyix.blogspot.com/2022/09/someones-been-messing-with-my-subnormals.html.  (RelStorage was mentioned there, but I'm pretty sure BTrees doesn't depend on RelStorage.)

I hadn't considered the possibility of GitHub Actions caches containing locally built persistent wheels.

**Comment by mgedmin** [comment]
I've wiped the Linux-pip-3.10 cache with

    gh api --method DELETE -H "Accept: application/vnd.github+json" /repos/zopefoundation/BTrees/actions/caches?key=Linux-pip-3.10

so let's see if that affects anything when the build is retried.

Incidentally, aarch64 builds take forever.  `.manylinux-install.sh` attempts to work around this by switching to -O1, but it prepends the -O1 flag at the front of CFLAGS, which means the -O3 from the environment specified in GitHub Workflow overrides it back to slow mode again.

**Comment by mgedmin** [comment]
Incidentally, I noticed that there are two workflows started with every push: one for the commit, and one for the PR.  The workflow skips all the tests for the PR workflow (yay avoiding duplication), but it still builds manylinux packages twice.  Bug in workflow configuration?

**Comment by mgedmin** [comment]
Looks like the cache clearing helped.

**Comment by mgedmin** [comment]
> Incidentally, aarch64 builds take forever. .manylinux-install.sh attempts to work around this by switching to -O1, but it prepends the -O1 flag at the front of CFLAGS, which means the -O3 from the environment specified in GitHub Workflow overrides it back to slow mode again.

Nah, I was wrong, there's an `export CFLAGS="-pipe"` before invoking .manylinux-install.sh that overrides the environment.  But even at -O1 the build takes over 2 hours!

**Comment by mgedmin** [comment]
Ok, now this looks good to me.

File: CHANGES.rst [hunk 1]
```diff
@@ -5,7 +5,8 @@
 4.10.1 (unreleased)
 ===================
 
-- Nothing changed yet.
+- Disable unsafe math optimizations in C code.
```
Comment (icemac) [inline CHANGES.rst:8]: @mgedmin Is this still correct?

Comment (mgedmin) [reply CHANGES.rst:8]: Yes

**Pull request merged**